sw/host/opentitantool/src/command/image.rs

```diff
@@ -160,77 +168,54 @@ impl CommandDispatch for ManifestUpdateCommand {
-        };
+        let ext = self
+            .manifest_ext
+            .as_deref()
+            .map(ManifestExtSpec::read_from_file)
+            .unwrap_or(Ok(Default::default()))?;
 
-        // Update `signed_region_end` if it is 0, otherwise it has already been set (i.e., to
-        // support offline signing scenarios).
-        if image.borrow_manifest()?.signed_region_end == 0 {
-            image.update_signed_region_end()?;
-        }
+        // Add any signed extensions defined in `ext`.
+        image.add_signed_manifest_extensions(&ext)?;
 
         let mut spx_private_key: Option<SpxKeypair> = None;
         if let Some(key) = &self.spx_key {
+            ext.check_for_duplicates(ManifestExtId::spx_key)?;
             let key = spx::load_spx_key(key)?;
-            // Add the SPX key extension
-            let key_bytes = key.pk_as_bytes();
-            buf.bytes[0..key_bytes.len()].copy_from_slice(key_bytes);
-            image.set_extension_entry(0, MANIFEST_EXT_ID_SPX_KEY, image.size as u32)?;
-            image.add_extension_data(
-                ManifestExtSpxKey {
-                    header: ManifestExtHeader {
-                        identifier: MANIFEST_EXT_ID_SPX_KEY,
-                        name: MANIFEST_EXT_NAME_SPX_KEY,
-                    },
-                    key: *LayoutVerified::<_, SigverifySpxKey>::new(&buf.bytes[0..key_bytes.len()])
-                        .unwrap()
-                        .into_ref(),
-                }
-                .as_bytes(),
-            )?;
+            let key_ext = ManifestExtEntry::new_spx_key_entry(&key)?;
+            image.add_manifest_extension(key_ext)?;
+
             if let SpxKey::Private(private) = key {
                 spx_private_key = Some(private);
             }
-
-            // Update `signed_region_end` after adding all signed extensions (which is just SPX key
-            // extension for now).
-            image.update_signed_region_end()?;
         }
 
-        if self.spx_key.is_some() {
-            image.set_extension_entry(1, MANIFEST_EXT_ID_SPX_SIGNATURE, image.size as u32)?;
-            image.add_extension_data(
-                ManifestExtSpxSignature {
-                    header: ManifestExtHeader {
-                        identifier: MANIFEST_EXT_ID_SPX_SIGNATURE,
-                        name: MANIFEST_EXT_NAME_SPX_SIGNATURE,
-                    },
-                    ..Default::default()
-                }
-                .as_bytes(),
+        // Determine if an SPX+ signature will be added and preallocate the extension.
+        if self.spx_signature.is_some() || spx_private_key.is_some() {
+            ext.check_for_duplicates(ManifestExtId::spx_signature)?;
+            image.allocate_manifest_extension(
+                ManifestExtId::spx_signature.into(),
+                std::mem::size_of::<ManifestExtSpxSignature>(),
             )?;
         }
 
         // Update `length` to be able to generate signatures.
```

Comment:
nit: can you mention that this is true by default and is only optional because we might want to test negative cases in which case length comes from the HJSON file.

Alternatively setting the length to the correct value unconditionally and overwriting it again only if the manifest spec includes it might be simpler.